### DISCUSSION TITLE
Generate Typescript types from schema

### DISCUSSION THREAD

**Discussion by karlismelderis-mckinsey** [authzed/spicedb#1872] [Q&A] [answered]
Has anyone attempted to generate TypeScript types from Spice schema?

```
schema: |-
  definition acme/user {}

  definition acme/system {
    relation admin: acme/user
    relation support: acme/user

    permission root_access = admin
    permission support_access = support
    permission create_something = admin
  }
```

would produce:
```ts
const acmeSystemRelations = {
  ADMIN: admin
  SUPPORT: support
} as const

type AcmeSystemRelations = typeof acmeSystemRelations[keyof typeof acmeSystemRelations]

const acmeSystemPermissions = {
  ROOT_ACCESS: 'root_access',
  SUPPORT_ACCESS: 'support_access',
  CREATE_SOMETHING: 'create_something',
} as const

type AcmeSystemPermissions = typeof acmeSystemPermissions[keyof typeof acmeSystemPermissions]
```

**Comment by josephschorr** [comment, marked as answer]
We actually have a prototype of a well-generated client for TypeScript, but we're still unsure of the best means of generating the TypeScript code necessary

**Reply by karlismelderis-mckinsey** [reply]
so I'm bit to early and have to wait 🫡
How soon do you feel anything will be available for trial and testing?

**Reply by vroldanbet** [reply]
There are currently no plans for this. Some community members have stepped up and created generator for other languages, like https://github.com/oviva-ag/spicegen

**Comment by Sefriol** [comment]
This could be done to almost any language with relatively little effort by using SpiceDB Reflect Schema API. However it does not have parsing information about he permissions themselves. Only relations.

This way you could just use dockerized spicedb instance, load the schema, use reflect API to generate necessary target language code.

For C# I was able to do for example:

```
RelationshipUpdateBuilder.Application().Admin().Group().AddRelationship();
```

Where Applications and Group are schema Definitions and Admin is a relation. You can do this with Permissions too, but unfortunately you cannot figure out potential Subject IDs and Types with it.

For example I could generate a function like this to make client code more nice:

```
Definitions.Application().Writer().CheckPermission("subjectTypeUser", "subjectId")
Definitions.Application().Writer().LookUpSubjects("subjectTypeUser", "subjectId")
Definitions.Application().Writer().LookUpResrouces("subjectTypeGroup", "subjectId")
```
But I kinda want something like this:

```
Definitions.Application().Writer().User().CheckPermission("subjectId")
Definitions.Application().Writer().Group().CheckPermission("subjectId")
Definitions.Application().Writer().User().LookUpSubjects("subjectId")
Definitions.Application().Writer().Group().LookUpResrouces("subjectId")
```

**Comment by hermannloose** [comment]
We built something for our internal use in https://github.com/yahiaosama/authz-schema-sync-check, that generates Typescript and Python types based on SpiceDB's tree-sitter grammar in https://github.com/authzed/tree-sitter-spicedb. It's built in Python and uses Jinja templates.

I think the most flexible approach might be for SpiceDB to provide a binary that does the generic parsing step and can invoke some popular templating engines (or just one reasonably popular one). Then it can ship with some opinionated templates for various languages (just types, not fully-fledged clients / DSLs) but users could always tweak it when necessary.

**Reply by verdverm** [reply]
They could have a command that emits the schema as JSON, and then we can read that into any template language. People have their particulars for the output when code gen'n after all

I've personally gone meta and write the schema in CUE and then generate both the spicedb schema and various language code to go with it, from the same source of truth. I like this better because my CUE schema can be far richer than what can be captured in a spicedb schema, allowing us to add more details for subsequent code gen tasks
